Mathlib/Analysis/Complex/Hadamard.lean

```diff
@@ -0,0 +1,487 @@
+    (hB : BddAbove ((abs ∘ f) '' (closedStrip 0 1))) : abs (f z) ≤ sSupAbsIm f (z.re) := by
+  refine le_csSup ?_ ?_
+  · apply BddAbove.mono (image_subset (abs ∘ f) _) hB
+    apply preimage_mono
+    simp only [singleton_subset_iff]
+    exact hD
+  · apply mem_image_of_mem (abs ∘ f)
+    simp only [mem_preimage, mem_singleton]
+
+lemma cocompact_strip.basis : HasBasis (cocompactStrip 0 1) (fun _ : ℝ × ℝ ↦ True )
+    (fun T ↦ (im ⁻¹' (Iic (T.1) ∪ Ici (T.2)) ∩ re ⁻¹' Icc 0 1)) := by
+  simp only [cocompactStrip, comap_principal, Real.cocompact_eq]
+  rw [← true_and True]
+  exact HasBasis.inf_principal (HasBasis.comap _ (HasBasis.sup atBot_basis atTop_basis)) _
+
+lemma cocompact_strip.mem_sets {s : Set ℂ} : s ∈ cocompactStrip 0 1 ↔ ∃ T, ∀ (z : ℂ),
+    z.re ∈ Icc 0 (1 : ℝ) → T ≤ |z.im| → z ∈ s := by
+  rw [HasBasis.mem_iff cocompact_strip.basis]
+  constructor
+  · rintro ⟨⟨a,b⟩, h⟩
+    simp only [preimage_union, ge_iff_le, zero_le_one, not_true, gt_iff_lt, true_and] at h
+    use max (-a) b
+    intros z hset him
+    apply h
+    refine mem_inter ?_ hset
+    simp only [mem_preimage, mem_Iic, mem_Ici, mem_union]
+    simp only [abs_ofReal, max_le_iff] at him
+    cases' him with hima himb
+    by_cases 0 ≤ z.im
+    · right; rwa [← _root_.abs_of_nonneg h]
+    · left; rwa [← neg_le_neg_iff, ← abs_of_neg (not_le.mp h)]
+  · rintro ⟨T, h⟩
+    use (-T, T)
+    simp only [preimage_union, true_and]
+    rintro z ⟨him, hset⟩
+    apply h _ hset
+    rwa [le_abs, ← le_neg, or_comm, ← mem_Iic, ← mem_Ici,
+      ← mem_union, ← mem_preimage, preimage_union]
+
+lemma _root_.Real.tendsto_sq_cocompact_atTop : Tendsto (fun x ↦ x^2) (cocompact ℝ) atTop := by
+  convert_to Tendsto (fun x ↦ ‖x‖ * ‖x‖) (cocompact ℝ) atTop
+  · ext x; rw [pow_two]; exact (abs_mul_abs_self _).symm
+  · exact tendsto_norm_cocompact_atTop.atTop_mul_atTop tendsto_norm_cocompact_atTop
+
+lemma cocompact_strip.tendsto_sq_im_atTop : Tendsto (fun z : ℂ ↦ z.im^2)
+    (cocompactStrip 0 1) atTop := by
+  apply tendsto_inf_left _
+  change Tendsto ((fun x ↦ x^2) ∘ im) (comap im (cocompact ℝ)) atTop
+  rw [tendsto_comap'_iff]
+  · convert_to Tendsto (fun x ↦ ‖x‖ * ‖x‖) (cocompact ℝ) atTop
+    · ext x; simp only [Real.rpow_two, pow_two, Real.norm_eq_abs, abs_mul_abs_self]
+    · exact tendsto_norm_cocompact_atTop.atTop_mul_atTop tendsto_norm_cocompact_atTop
+  ·  simp only [univ_mem, range_im]
+
+lemma cocompact_strip.eventually_sq_re_le_one : ∀ᶠ (z : ℂ) in
+    (cocompactStrip 0 1), z.re^2 ≤ 1 := by
+  rw [cocompactStrip, comap_principal, eventually_inf_principal]
+  apply eventually_of_forall
+  rintro z hz
+  simp only [Real.rpow_two]
+  rw [sq_le_one_iff hz.1]
+  exact hz.2
+
+-- Smoothly decreasing function when the real part is bounded eventually ≤ 1
+lemma expterm_eventually_le_one (C : ℝ) (n : ℕ) (hn : 1 ≤ n) : ∀ᶠ (z : ℂ)
+    in (comap Complex.im (cocompact ℝ) ⊓ (comap Complex.re (principal  (Icc 0 1)))),
+    C * abs (exp ((z^(2 : ℕ)-1) * (n : ℝ)⁻¹) ) ≤ 1 := by
```

Comment:
Can you use vertical lines instead of `abs`, to make things more readable? Also, spurious space before your last parenthesis.

Reply:
If that is ok, I would prefer keeping `abs`. I found that the vertical lines would just get lost in the parentheses.

Reply:
You should be able to drop at least one set:
```suggestion
    C * abs (exp ((z^2-1) * (n : ℝ)⁻¹) ) ≤ 1 := by
```

Reply:
This actually breaks for me.